File: voidshare/p2p_encryption/p2p/peer.ts
import * as fs from 'fs';
import * as net from 'net';
import { AES256 } from '../encryption/aes';
import { RSAKeyExchange } from '../encryption/rsa';
import { DigitalSignature } from '../encryption/signature';

export class Peer {
  private rsa: RSAKeyExchange;
  public publicKey: string;
  public privateKey: string;

  constructor(private port: number = 5000, private host: string = 'localhost') {
    this.rsa = new RSAKeyExchange();
    const keys = this.rsa.exportKeys();
    this.publicKey = keys.publicKey;
    this.privateKey = keys.privateKey;
  }

  sendFile(filePath: string, receiverHost: string, receiverPort: number): void {
    const sessionKey = crypto.randomBytes(32);
    const aes = new AES256(sessionKey);
    const encryptedFile = filePath + '.enc';

    aes.encrypt(filePath, encryptedFile);

    const receiverPublicKey = fs.readFileSync('receiver_pub.pem', 'utf-8');
    const encryptedSessionKey = this.rsa.encryptSessionKey(sessionKey, receiverPublicKey);
    const signature = DigitalSignature.signFile(encryptedFile, this.privateKey);

    const fileBuffer = fs.readFileSync(encryptedFile);

    const client = new net.Socket();
    client.connect(receiverPort, receiverHost, () => {
      const delimiter = Buffer.from('||');
      client.write(Buffer.concat([encryptedSessionKey, delimiter, signature, delimiter, fileBuffer]));
      client.end();
    });
  }

  receiveFile(savePath: string): void {
    const server = net.createServer((socket) => {
      const chunks: Buffer[] = [];

      socket.on('data', (data) => chunks.push(data));

      socket.on('end', () => {
        const receivedData = Buffer.concat(chunks);
        const parts = receivedData.toString().split('||');

        const encryptedSessionKey = Buffer.from(parts[0], 'binary');
        const signature = Buffer.from(parts[1], 'binary');
        const fileContent = Buffer.from(parts[2], 'binary');

        const encryptedFilePath = savePath + '.enc';
        fs.writeFileSync(encryptedFilePath, fileContent);

        const sessionKey = this.rsa.decryptSessionKey(encryptedSessionKey);

        const senderPublicKey = fs.readFileSync('sender_pub.pem', 'utf-8');
        const isValid = DigitalSignature.verifySignature(encryptedFilePath, signature, senderPublicKey);

        if (!isValid) {
          console.error('Signature verification failed!');
          return;
        }

        const aes = new AES256(sessionKey);
        aes.decrypt(encryptedFilePath, savePath);
        console.log('File received and decrypted successfully!');
      });
    });

    server.listen(this.port, this.host);
  }
}

File: voidshare/p2p_encryption/p2p/peer.py
import socket
import os
from encryption.aes import AES256
from encryption.rsa import RSAKeyExchange
from encryption.signature import DigitalSignature

class Peer:
    def __init__(self, host='localhost', port=5000):
        self.host = host
        self.port = port
        self.rsa = RSAKeyExchange()
        self.public_key, self.private_key = self.rsa.export_keys()

    def send_file(self, file_path, receiver_host, receiver_port):
        # Generate AES session key
        session_key = os.urandom(32)

        # Encrypt the file using AES
        encrypted_file = file_path + ".enc"
        aes = AES256(session_key)
        aes.encrypt(file_path, encrypted_file)

        # Encrypt AES session key using RSA
        with open("receiver_pub.pem", "rb") as f:
            receiver_public_key = f.read()
        encrypted_session_key = self.rsa.encrypt_session_key(session_key, receiver_public_key)

        # Sign the encrypted file
        signature = DigitalSignature.sign_file(encrypted_file, self.private_key)

        # Send file
        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
            s.connect((receiver_host, receiver_port))
            s.sendall(encrypted_session_key + b"||" + signature)

            with open(encrypted_file, "rb") as f:
                s.sendall(f.read())

    def receive_file(self, save_path):
        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
            s.bind((self.host, self.port))
            s.listen(1)
            conn, _ = s.accept()
            with conn:
                data = conn.recv(2048)
                encrypted_session_key, signature = data.split(b"||", 1)

                received_file_path = save_path + ".enc"
                with open(received_file_path, "wb") as f:
                    f.write(conn.recv(1024 * 1024))  # Receive encrypted file

                # Decrypt session key
                session_key = self.rsa.decrypt_session_key(encrypted_session_key)

                # Verify signature
                with open("sender_pub.pem", "rb") as f:
                    sender_public_key = f.read()
                if not DigitalSignature.verify_signature(received_file_path, signature, sender_public_key):
                    print("Signature verification failed!")
                    return

                # Decrypt the file
                aes = AES256(session_key)
                aes.decrypt(received_file_path, save_path)
                print("File received and decrypted successfully!")
